### PR TITLE
Complete `DefaultClientRequestContext.whenIntialized()` after fully initializing a context

### LINKED ISSUE #3381: Intermittent test failure: LazyDynamicEndpointGroupTest.emptyEndpoint()
```
LazyDynamicEndpointGroupTest > emptyEndpoint() FAILED
    java.lang.AssertionError:
    Expecting a throwable with cause being an instance of:
     <com.linecorp.armeria.client.UnprocessedRequestException>
    but was an instance of:
     <com.linecorp.armeria.common.stream.ClosedStreamException>
        at com.linecorp.armeria.client.grpc.LazyDynamicEndpointGroupTest.emptyEndpoint(LazyDynamicEndpointGroupTest.java:91)
```

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`LazyDynamicEndpointGroupTest.emptyEndpoint()` sometimes failed in CI
builds. #3381 It expects to fail the test with `EmptyEndpointException`,
however, it fails with `ClosedStreamException`.

After digging the issue, I find out that there is a race in `whenInitialized`.
If the `acquiredEventLoop.execute()` is executed immediately and completed earlier than
returning the method,
https://github.com/line/armeria/blob/d4880fe12690d2dafd2c5e7fa9f24c3b24837a00/core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java#L307-L309
the callbacks of `ctx.whenInitialized()` will be invoked before a `RequestLog`
is completed.
https://github.com/line/armeria/blob/207c5e038f59802dca769936a50e219a5fe308ea/grpc/src/main/java/com/linecorp/armeria/internal/client/grpc/ArmeriaClientCall.java#L337-L348
As the `req` is closed already, the `req.write()` would be failed with
`ClosedStreamException`.

Modifications:

- Complete `DefaultClientRequestContext.whenIntialized()` after
  `initContextAndExecuteWithFallback` of `ClientUtil`

Result:

- You no longer see `ClosedStreamException` when an `EndpointGroup` is empty.
- Fixes #3381

### CODE REVIEW

**Review by minwoox — APPROVED** [review]
What a brilliant!

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java [hunk 1]
```diff
@@ -329,6 +310,21 @@ private static ServiceRequestContext serviceRequestContext() {
         }
     }
 
+    /**
+     * Completes the {@link #whenInitialized()} with the specified value.
+     */
+    public void whenInitialized(boolean success) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:316]: How about renaming to `finishInitialization()`?

**Review by trustin — COMMENTED** [review]
Nice find!

File: core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:316]: That sounds better. 👍

**Review by trustin — APPROVED** [review]
Thanks, @ikhoon !

**Pull request merged**